src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+#include "catalog/indexing.h"
+#include "catalog/pg_depend.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+
+#include "distributed/dist_catalog/dependency.h"
+#include "distributed/dist_catalog/distobject.h"
+
+
+static bool SupportedDependencyByCitus(const ObjectAddress *address);
+static bool IsObjectAddressInList(const ObjectAddress *findAddress, List *addressList);
+static bool IsObjectAddressOwnedByExtension(const ObjectAddress *target);
+
+static void recurse_pg_depend(const ObjectAddress *target,
+							  bool (*follow)(void *context, const Form_pg_depend row),
+							  void (*apply)(void *context, const Form_pg_depend row),
+							  void *context);
+static bool follow_order_object_address(void *context, const Form_pg_depend pg_depend);
+static bool follow_get_dependencies_for_object(void *context,
+											   const Form_pg_depend pg_depend);
+static void apply_add_to_target_list(void *context, const Form_pg_depend pg_depend);
+
+
+/*
+ * GetDependenciesForObject returns a list of ObjectAddesses to be created in order
+ * before the target object could safely be created on a worker. Some of the object might
+ * already be created on a worker. It should be created in an idempotent way.
+ */
+List *
+GetDependenciesForObject(const ObjectAddress *target)
+{
+	List *dependencyList = NIL;
+	recurse_pg_depend(target,
+					  &follow_get_dependencies_for_object,
+					  &apply_add_to_target_list,
+					  &dependencyList);
+	return dependencyList;
+}
+
+
+/*
+ * IsObjectAddressInList is a helper function that can check if an ObjectAddress is
+ * already in a (unsorted) list of ObjectAddresses
+ */
+static bool
+IsObjectAddressInList(const ObjectAddress *findAddress, List *addressList)
+{
+	ListCell *addressCell = NULL;
+	foreach(addressCell, addressList)
+	{
+		ObjectAddress *currentAddress = (ObjectAddress *) lfirst(addressCell);
+
+		/* equality check according as used in postgres for object_address_present */
+		if (findAddress->classId == currentAddress->classId && findAddress->objectId ==
+			currentAddress->objectId)
+		{
+			if (findAddress->objectSubId == currentAddress->objectSubId ||
+				currentAddress->objectSubId == 0)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * SupportedDependencyByCitus returns whether citus has support to distribute the object
+ * addressed.
+ */
+static bool
+SupportedDependencyByCitus(const ObjectAddress *address)
+{
+	/*
+	 * looking at the type of a object to see if we know how to create the object on the
+	 * workers.
+	 */
+	switch (getObjectClass(address))
+	{
+		case OCLASS_SCHEMA:
+		{
+			return true;
+		}
+
+		default:
+		{
+			/* unsupported type */
+			return false;
+		}
+	}
+
+	/*
+	 * all types should have returned above, compilers complaining about a non return path
+	 * indicate a bug in the above switch. Missing returns or breaking code flow should be
+	 * addressed in the switch statement, preferably not here.
+	 */
```

Comment:
I think this will be permanently addressed by the `default:`, maybe redundant comment.